### ISSUE TITLE
Schema: $graveyardHolder is not registered.

### ISSUE THREAD

**Issue by scofalik** [ckeditor/ckeditor5-engine#638]
**TLDR:**
`LiveSelection` sometimes end up in graveyard. Even if it is corrected a few lines after, it still triggers errors. One of those errors is checking `LiveSelection` position with `Schema`. Then, it throws `model-schema-no-item: Item with specified name does not exist in schema.`.

We can either fix the issue directly by adding `$graveyardHolder` to `Schema` or fix the `LiveSelection` behavior. Or implement one of other proposed solutions.

This can be reproduced by having a one paragraph in editor, selecting all contents and clicking delete.

**Long version:**

Element with name `$graveyardHolder` is not registered in `Schema`. This means that checking anything in graveyard results in an error: `model-schema-no-item: Item with specified name does not exist in schema.`.

Normally, there should not be many cases where we need to check if something in graveyard is allowed by `Schema`. The problem appears when `LiveSelection` is moved to a graveyard. Then, some features (like `Link` right now) may check if command should be enabled by checking `Selection` position with `Schema`.

This is a bit "incorrect scenario" because document selection should not end up in graveyard. Features (or maybe even selection itself) should take care of it. Nevertheless, right now it does end up for a short while, i.e. when editor has one paragraph and it is wholly selected and removed. The selection is quickly restored in the document, but for a while it is in graveyard.

There are four solutions:

1.  We fix it in `(Live)Selection`. I think that some time ago we agreed that this kind of situation could be handled by `LiveSelection` itself. This is the most time-consuming selection, but with the new events handling, introduced in recent PRs, it should not be that difficult. I'd go with it. Read below if you want to hear alternatives.

1. Add `$graveyardHolder` to `Schema`. This soultion does not care that graveyard is not a correct place for document selection. It fixes the direct root of the problem. I can get behind this solution but it doesn't look at problem in a big picture.

2. The problem is strictly connected with deleting nodes from model. Though features may use `batch.remove()` and trigger this error, right now the only source of the problem is `deleteContents`. We could try to fix it there and keep fingers crossed that the problem won't appear again. I don't think this is a good solution because we might run into it again and again and again.

3.  Features should check if document selection is in graveyard, and set some default/failsafe values if it is there. The drawback here is that each developer has to acknowledge this scenario when creating their feature.

**Comment by Reinmar** [comment]
I've got the same problem with the clipboard pipeline. I need to process data in a forgiving context, namely – something which works like a `$root`, but allows also inline stuff inside it (like `$text`).

So I could do something like:

```
schema.registerItem( '$clipboardHolder', '$root' );
schema.allow( '$text', { inside: '$clipboardHolder' } );
```

This may work fine (I haven't tested it yet). I have doubts though, about some edge cases like:

* What if `<x>` can only be inside `<y>` but the pasted content was incomplete? Theoretically, it shouldn't be, so we should be safe.
* What if someone has two roots with different configurations. Processing in context of the main one will be wrong. But this again should work – we simply need to use the correct context during processing.
* What if `<x>` can be only inside a `<tableCell>`, not inside `<$root>`? Then, we'll filter it out since `$clipboardHolder` inherited from `$root` and not `tableCell`...

And here problem begins. We're not able to tell from the clipboard perspective whether the context should be inherited from the `tableCell` or its parent root. There are two options I can think of:

1. We need to mark some items in the schema as "limitters" (something similar to CKE4's `blockLimit`). If a selection is inside a limitter, that limitter should be used instead of the root. So it's a kind of virtual root. Still, based on such a virtual root, we'll need to create on the fly a schema item which inherits from it and allows text.
2. We process the clipboard content in "allow all" schema tree. The correctness of the structure will be ensured later down the pipeline, inside `Composer.insertData()`. Note that the composer's method will have to check the schema anyway, because pasting a paragraph into a list item needs to filter the paragraph out anyway.

Based on option 2 (which I quite like) I thought about one more thing – what if pasted content was processed precisely in the element in which selection is currently placed? That'd give us automatic "filter out" mechanism, but unfortunately, it'd fail in couple of cases. E.g. what if a widget is selected (so the anchor is in the `$root`) or if the entire content was selected (and selection was anchored in the `$root`)? A text that was pasted and should be autoparagraphed would be lost again.

We'll need a brainstorm I'm afraid.

**Comment by Reinmar** [comment]
> Based on option 2 (which I quite like) I thought about one more thing – what if pasted content was processed precisely in the element in which selection is currently placed? That'd give us automatic "filter out" mechanism, but unfortunately, it'd fail in couple of cases. E.g. what if a widget is selected (so the anchor is in the $root) or if the entire content was selected (and selection was anchored in the $root)? A text that was pasted and should be autoparagraphed would be lost again.

Actually, it's totally stupid :D. The current element cannot be the context, because then pasting a list inside a paragraph wouldn't work at all. The context is still the root. So the first option, about "limiters" sounds better now :D.

**Comment by Reinmar** [comment]
> 2. We process the clipboard content in "allow all" schema tree. The correctness of the structure will be ensured later down the pipeline, inside Composer.insertData(). Note that the composer's method will have to check the schema anyway, because pasting a paragraph into a list item needs to filter the paragraph out anyway.

The problem with this option is that we'd duplicate the "filter out" mechanism of the conversion inside the `insertData()` method. I wonder if that's a problem or not and it's a question to @scofalik and @pjasiun – how good is the filtering which converters provide? Is it something that should do us 90% of the job, or it's just a useful touch, but the content must later be checked anyway?

**Comment by scofalik** [comment]
It appears that this is a very serious regression. The error is thrown each time a selection range is removed (selection starts and ends in same paragraph).

The only reason you haven't witnessed it yet is because it shows up only if `List` feature is enabled.

**Comment by scofalik** [comment]
Related: https://github.com/ckeditor/ckeditor5-list/issues/16

**Comment by scofalik** [comment]
Okay, it becomes more and more obvious that `LiveSelection` should not end up in graveyard. Here's next case where it is problematic:

1. Consider collaborative editing between two clients.
2. First client have `<p>foo^bar</p>`.
3. Second client removes `oba`.
4. Document selection ends in graveyard.

**Comment by scofalik** [comment]
DUP #331
